src/test/java/api/CirculationRulesEngineAPITests.java

```diff
@@ -387,20 +387,9 @@ void cachedRulesAreUsedEvenWhenRulesInStorageHaveBeenChanged() {
   }
 
   @Test
-  void cacheIsInvalidatedAfterFiveSeconds() {
-    setRules(rulesFallback);
-    assertThat(applyRulesForLoanPolicy(m1, t1, g1, s1), is(lp6));
-
-    circulationRulesFixture.updateCirculationRulesWithoutInvalidatingCache(
-      rulesFallback2);
-
-    // Poll until the cached rules should have been replaced
-    await()
-      .atLeast(4, SECONDS)
-      .atMost(6, SECONDS)
-      .pollDelay(1, SECONDS)
-      .pollInterval(1, SECONDS)
-      .until(() -> applyRulesForLoanPolicy(m1, t1, g1, s1), is(lp7));
+  void canRefreshRules(){
```

Comment:
What is this test intended to demonstrate?

If it is to check that the rules have been reloaded, I might want an assertion of some sort to check that new rules were in place after the reload.

Reply:
It demonstrates that when there are stored rules in place, invoking the reload endpoint gets the appropriate response.

There is no way to test that the reload endpoint reloads the rules from within an automated test.  Or at least, none I can see.  When testing this locally, I establish this by watching the logs and timing circulation operations.  The code block that reloads the rules has a lot of logging so I can see an invocation of the endpoint kicks off a reload.  The cache object also tracks the last time it was updated and logs that.  I can also time circ operations and verify that they are faster because they are not triggering a reload (again, by looking at the logs).  I don't know of any way to do any of that from within an automated test.  If you do, please share.

Reply:
It demonstrates that when there are stored rules to reload, an invocation of the endpoint returns the appropriate response.

I cannot think of any way to test that the reload endpoint actually reloads the rules from within an automated test. Remember that any attempt to retrieve the rules will aslo trigger a rebuild if they have not been built. I verify this when testing locally by watching the logs and timing circ operations, which are measurably faster when not rebuilding the rules. So I can trigger a reload via the endpoint, watch the log to see that the rebuild code is triggering, then try a circ operation, and watch the logfile to see that one is not being triggered, and that the operation itself is faster, which is the desired and expected behavior. I don't know how to do any of that within an automated test. Do you? If so, please share.

Reply:
> I cannot think of any way to test that the reload endpoint actually reloads the rules from within an automated test. Remember that any attempt to retrieve the rules will also trigger a rebuild if they have not been built.

> I don't know how to do any of that within an automated test. Do you?

Not by monitoring the logs. It is possible to do that from tests, however it is clunky and brittle and I wouldn't advise it.

I would likely use an approach similar to the test that was removed during the is change:
* Set the rules with one definition
* Change the rules in storage
* Trigger the reload endpoint
* Apply the rules to an example confirming that the new rules are now being used

If the endpoint doesn't respond till the rules have been reloaded, then there may not be a need for awaitility code to poll the application of the rules.

Yes, it is still possible that the behind the scenes fallback does the reloading. To me, that's still more useful than a test that only checks a response, which, to me, is primarily a way to boost coverage on the code that implements that endpoint.

Reply:
I don't see how that's a more valid test of the endpoint, given that you can't isolate what operation is actually rebuilding the rules.  However, it's pretty easy to write, and basically includes the original test, so it's not a hill I'm interested in dying on.  I'll modify the test.

Reply:
This was purely a suggestion. I'm not responsible for this module or work, and you are closer to the design, so if you don't think it's valuable, don't do it

Reply:
I may have misunderstood the design. I thought the rules only automatically reloaded when they weren't set

If they somehow automatically refresh when storage is changed (in production code, not tests) then I don't understand why the timer endpoint would be needed at all

If the automatic refresh when storage changes is a test only behaviour, then I'd changed the way the fixtures used in the tests behave to only do a reload after storage changed via the timer endpoint. That way, it'll be tested naturally in a bunch of tests

Reply:
> Check the code that does a put on the circulation rules.

I don't have capacity to do a review to that level of detail, sorry.

Would you like me to stop responding, close this comment and let others review it instead?

> The first thing the reload code does is pull the rules in storage, so they have to be the updated rules. But the put endpoint would have to wait to send the response until the reload returns a result, or that causes test failures because tests expect the rules to be rebuilt immediately when a put is done

I'm definitely not following along. If the PUT endpoint is production code, why couldn't it wait till the reload of the rules is done before responding?

Similarly, if it's test code, change it to suit the new way the system works.



> So I'm assuming there's some reason I don't know? Maybe one of the other reviewers from Vega can speak to that.

I'm afraid I can't help with the reason these changes are being made or the motivation to use a timer endpoint.

Reply:
@K-Felk Ok, I'll pick it up from here. Updating the rules during PUT seems to be the best solution as it'll help to avoid the delay during the first post-PUT action that requires rules (like check-out). If it causes some test failures, maybe there's an issue with tests?

Reply:
The issue is that the reload:
1.  Can only start once the initial put is successful, and should only run if and when it's successful
2.  Has to be complete before the response from the put is written out

After hours of trying and asking for help, I have been unable to write code that does this.  At best, I can trigger the reload after the put, but the code does not wait for the reload to be successful before writing the result, leading to intermittent test failures, because tests expect to be able to access the rules *immediately* once the put is done.  Changing that expectation would involve changing all the tests to incorporate some kind of delay, which is what I'm trying to avoid.  In actual production, waiting a few seconds after put to reload the rules is likely NBD, because you still have the old rules cached until the reload is complete, but it messes up automated tests badly.

Current configuration works because circ operations will wait for the rebuild to complete before finishing if the cache is empty.

Reply:
@K-Felk 

> After hours of trying and asking for help, I have been unable to write code that does this.

I'm sorry I wasn't better help earlier. I've put together a [branch](https://github.com/folio-org/mod-circulation/compare/CIRC-1783...CIRC-1783-marc) (based upon this work) that reloads the rules when new definitions are submitted.

> the code does not wait for the reload to be successful before writing the result, leading to intermittent test failures, because tests expect to be able to access the changed rules immediately once the put is done

I've run the build once and the tests passed. However, I don't know if the intermittent failures you encountered might be present in the work I've done too.